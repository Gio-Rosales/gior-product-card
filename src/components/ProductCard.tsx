import React, { createContext } from 'react';
import { useProduct } from '../hooks/useProduct';

import { ProductContextProps, ProductCardProps } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = (
  {
    product,
    children,
    className,
    style,
    onChange,
    value,
    initialValues
  }: ProductCardProps) => {
  const { counter, isMaxCountReached, maxCount, increaseBy, reset } = useProduct({ onChange, product, value, initialValues });

  // console.log('counter', counter);
  return (
    <Provider value={{
      product,
      counter,
      increaseBy,
      maxCount
    }}>
      <div className={`${styles.productCard} ${className}`} id={product.id} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,

          increaseBy,
          reset
        })}
      </div>
    </Provider>
  )
};
