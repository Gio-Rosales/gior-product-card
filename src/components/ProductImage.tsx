import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';

import { ProductImageProps } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export const ProductImage = ({ img, className, style }: ProductImageProps) => {
  const { product } = useContext(ProductContext);
  let imageToShow: string;

  if (img) {
    imageToShow = img;
  } else if (product.img) {
    imageToShow = product.img;
  } else {
    imageToShow = noImage;
  }

  return (
    <img
      src={imageToShow}
      alt="Product"
      className={`${styles.productImg} ${className}`}
      style={style}
    />
  );
};