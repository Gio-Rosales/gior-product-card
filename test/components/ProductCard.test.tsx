import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { productWOImage } from '../data/products';

const {act} = renderer;

describe('ProductCard', () => {
  it('should show correctly', () => {
    const wrapper = renderer.create(
      <ProductCard product={productWOImage}>
        {
          () => (
            <h1>Hola</h1>
          )
        }
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('shoudl increment the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={productWOImage} >
        {
          ({count, increaseBy}) => (
            <>
              <h1>Product Card</h1>
              <span>{count}</span>
              <button onClick={() => increaseBy(1)}>1</button>
            </>
          )
        }
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      (tree as any).children[2].props.onClick();
    })

    tree = wrapper.toJSON();

    console.log((tree as any).children[1].children);
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
