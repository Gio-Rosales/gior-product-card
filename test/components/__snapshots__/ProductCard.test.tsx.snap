// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard shoudl increment the counter 1`] = `
<div
  className="productCard undefined"
  id="1"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    1
  </button>
</div>
`;

exports[`ProductCard should show correctly 1`] = `
<div
  className="productCard undefined"
  id="1"
>
  <h1>
    Hola
  </h1>
</div>
`;
